compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -535,7 +535,12 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                         }
                     }
                 }
-
+                ty::Adt(def, _) => {
+                    let ty = def.sized_constraint(self.tcx()).0;
```

Comment:
This still breaks if you changed the definitions from:

```
struct A<T>(B<T>);
//~^ ERROR: recursive types `A` and `B` have infinite size
struct B<T>(A<A<T>>);
```

to 

```
struct A<T>(B<T>, ());
//~^ ERROR: recursive types `A` and `B` have infinite size
struct B<T>(A<A<T>>, ());
```

Reply:
Nevermind, I've misinterpreted why you're using `sized_constraint` here.

Checking that `sized_constraint` matches `[ty::Error]` doesn't seem like the right way to check that the struct is cyclical, and after that, not registering an `AutoImplCandidate` if a type is cyclical also seems like the wrong way to avoid an overflow in the solver.

The correct solution is probably to not ICE on the overflow at all, and instead treat it as ambiguity, but that is (as I said above) a bit delicate...

Reply:
If we make the following changes:
```
struct A<T>(B<T>);
struct B<T>(A<T>);
```
it will not cause a panic because the code has been checked for cycles in the ﻿`check_evaluation_cycle` function (see [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_trait_selection/src/traits/select/mod.rs#L1049-L1051)).
Therefore, could we add a similar function called ﻿`check_evaluation_overflow` to check if the current stack has an infinite evaluation process?

Reply:
I don't think that that's right either. I think we should probably just make `predicate_may_hold_fatal` not actually ICE. But then there's only one callsite for `predicate_may_hold_fatal` anyways, so it's probably best to just inline it and delay a bug if we have overflow.